tests/abi/functions.h

```diff
@@ -0,0 +1,273 @@
+                          int16_t _02,
+                          uint8_t _03,
+                          __int128_t _04) {
+  PRINT_HEADER(multitypeArgumentFunction);
+  puts("  Arguments:");
+  PRINT_INT64_T(_00);
+  PRINT_UINT32_T(_01);
+  PRINT_INT16_T(_02);
+  PRINT_UINT8_T(_03);
+  PRINT_INT128_T(_04);
+  puts("  ReturnValues:");
+  puts("");
+}
+
+void __attribute__((noinline, weak)) ABIDEF
+functionForRandomnessTesting(uint8_t _00,
+                             uint8_t _01,
+                             uint8_t _02,
+                             uint8_t _03,
+                             uint8_t _04,
+                             uint8_t _05,
+                             uint8_t _06,
+                             uint8_t _07,
+                             uint8_t _08,
+                             uint8_t _09,
+                             uint8_t _10,
+                             uint8_t _11,
+                             uint8_t _12,
+                             uint8_t _13,
+                             uint8_t _14,
+                             uint8_t _15,
+                             uint8_t _16) {
+  PRINT_HEADER(functionForRandomnessTesting);
+  puts("  Arguments:");
+  PRINT_UINT8_T(_00);
+  PRINT_UINT8_T(_01);
+  PRINT_UINT8_T(_02);
+  PRINT_UINT8_T(_03);
+  PRINT_UINT8_T(_04);
+  PRINT_UINT8_T(_05);
+  PRINT_UINT8_T(_06);
+  PRINT_UINT8_T(_07);
+  PRINT_UINT8_T(_08);
+  PRINT_UINT8_T(_09);
+  PRINT_UINT8_T(_10);
+  PRINT_UINT8_T(_11);
+  PRINT_UINT8_T(_12);
+  PRINT_UINT8_T(_13);
+  PRINT_UINT8_T(_14);
+  PRINT_UINT8_T(_15);
+  PRINT_UINT8_T(_16);
+  puts("  ReturnValues:");
+  puts("");
+}
+
+void __attribute__((noinline, weak)) ABIDEF single16bitArgument(uint16_t _00) {
+  PRINT_HEADER(single16bitArgument);
+  puts("  Arguments:");
+  PRINT_UINT16_T(_00);
+  puts("  ReturnValues:");
+  puts("");
+}
+
+void __attribute__((noinline, weak)) ABIDEF single32bitArgument(int32_t _00) {
+  PRINT_HEADER(single32bitArgument);
+  puts("  Arguments:");
+  PRINT_INT32_T(_00);
+  puts("  ReturnValues:");
+  puts("");
+}
+
+void __attribute__((noinline, weak)) ABIDEF single64bitArgument(uint64_t _00) {
+  PRINT_HEADER(single64bitArgument);
+  puts("  Arguments:");
+  PRINT_UINT64_T(_00);
+  puts("  ReturnValues:");
+  puts("");
+}
+
+void __attribute__((noinline, weak)) ABIDEF
+single128bitArgument(__uint128_t _00) {
+  PRINT_HEADER(single128bitArgument);
+  puts("  Arguments:");
+  PRINT_UINT128_T(_00);
+  puts("  ReturnValues:");
+  puts("");
+}
+
+/* clang-format off */
+#define ARGUMENT_FUNCTION_LIST functionWithLotsOfArguments,  \
+                               multitypeArgumentFunction,    \
+                               functionForRandomnessTesting, \
+                               single16bitArgument,          \
+                               single32bitArgument,          \
+                               single64bitArgument,          \
+                               single128bitArgument
```

Comment:
Let's try to keep macro magic as local as possible.
```c
typedef struct {
  const char *name;
  void *function;
} test_function_t;

#define ENTRY(name) { #name, name }

test_function_t test_functions[] = {
  ENTRY(functionWithLotsOfArguments),
};

#undef ENTRY
```
And then you can iterate on `test_functions` without macro magic.

Reply:
An `IterationOn` function needs to be generated for each of those, so I wasn't able to do anything about that. Let's solve the [relevant discussion](https://github.com/revng/revng-qa/pull/4#discussion_r763003454) first - that should give an answer to this problem as well.

Reply:
So, to recap this, using runtime function pointers is painful, I'd rather let linker take care of the converting function names to addresses - hence the compilation time loops.